### ISSUE TITLE
[12.0][stock] Not possible to unreserve more products than you have in stock

### ISSUE THREAD

**Issue by JordiBForgeFlow** [odoo/odoo#62139]
Impacted versions: 12.0

Steps to reproduce:
It's not possible to reproduce in normal circumstances. We have faced this error in situations of multiple concurrent users trying to pick.

Picking 1 is created, reserving stock in location A.
Picking 2 is created, reserving stock in location A.
Inventory adjustment is created, updating quantity in location A

User 1 and 2 complete the picking concurrently, and the inventory adjustment may be done during the process.

Current behavior: 

A mismatch between the reserved quantity in quants and in move lines results in the following error: https://github.com/odoo/odoo/blob/12.0/addons/stock/models/stock_quant.py#L269 when the user attempts to complete the picking.

We also find cases where there are move lines with reserved quantity, but there's no quant.

Expected behavior: The integrity between quants and move lines should be always respected.

Have you faced this inconsistency between quants and move lines before @amoyaux @rco-odoo? It's hitting us pretty bad. 

We had to end up designing a con job that runs every 10 minutes to locate the inconsistent quant/move lines and align them. Not good.

**Comment by amoyaux** [comment]
@JordiBForgeFlow Thnaks to report it. We gonna check for it

**Comment by yelizariev** [comment]
opw-2388684

**Comment by JordiBForgeFlow** [comment]
@yelizariev where do you get the opw code? Is that the support ticket number?

**Comment by yelizariev** [comment]
Yes, it's your support ticket number.
I thought to work on it myself, then found that @amoyaux took it

**Comment by Yenthe666** [comment]
By the way if you make a Github ticket and an OPW then do link them together so both places are aware of eachother :) Otherwise we're possible wasting Odoo it's resources.

**Comment by yelizariev** [comment]
Can it be because of this code?

https://github.com/odoo/odoo/blob/95d28dc168b95f9a8f5a95b67b4b41bc3af1714d/addons/stock/models/stock_quant.py#L217-L229

as I understand, the idea is that if multiple process call ``_update_available_quantity``, then only one will make update for specific quant. But that lock can be acquired by another code.

@amoyaux may it be the reason? If so, 

1. can a workaround be removing ``NOWAIT``? 
2. What could be a solution? Making a new lock specifically for this method, .e.g ``pg_advisory_lock``?

**Comment by amoyaux** [comment]
@yelizariev I think this code is correct. I would rather bet on the _update_reserved_quantity that maybe does not support concurrency very well and could create a mistake if a savepoint doesn't rollback the move line information

**Comment by yelizariev** [comment]
@amoyaux I don't agree. I checked how NOWAIT works:

1. add a sleep to stock.picking writing
```
    @api.multi
    def write(self, vals):
        res = super(Picking, self).write(vals)
        import time; time.sleep(10)
```
2. make update via UI
3. while it's loading try to get the lock:
```
devel=# SELECT 1 FROM stock_picking WHERE id = 30 FOR UPDATE NOWAIT;
ERROR:  could not obtain lock on row in relation "stock_picking"
```
4. Once timer is over it works:

```
devel=# SELECT 1 FROM stock_picking WHERE id = 30 FOR UPDATE NOWAIT;
 ?column? 
----------
        1
(1 row)
```

**Comment by Julien00859** [comment]
Hello, just to mention that a `FOR UPDATE` kind of lock is set by the operations `INSERT`, `UPDATE`, `DELETE` and `SELECT ... FOR UPDATE`. Since you overwrite `write` to sleep *after*  the orm's write (which do an `UPDATE` statement) but *before* the orm commits the transaction, there the written rows are locked.

**Comment by JordiBForgeFlow** [comment]
We actually get a lot of "could not obtain lock on row in relation "stock.quant"" in the locks.

**Comment by yelizariev** [comment]
@JordiBForgeFlow in sql logs? It's not a problem by itself. The problem happens when the lock is acquired outside of ``_update_available_quantity``

**Comment by yelizariev** [comment]
I still think that removing NOWAIT is a workaround (may work slowly because of retries after serialization failure, but will keep data consistent). But I'm not sure for a fix. I'll check tomorrow

**Comment by yelizariev** [comment]
What about removing NOWAIT only for inventory adjustment?
https://github.com/odoo/odoo/pull/62515

I found a way to reproduce potentially related problem

**Comment by randall-vx** [comment]
@JulioSerna @moylop260 steps to reproduce

**Comment by gurneyalex** [comment]
@JordiBForgeFlow until this is fixed can you share the code of your cron ?

**Comment by amoyaux** [comment]
@JordiBForgeFlow We checked and were not able to reproduce it at all. Do you have more details or something that could help us.
Here is a test that we used to check the robustness of concurrency. Could you try it on your side, and let me know the result.
:warning: **Do not use it on your production database.** It doesn't use a test cursor and environment, so it could break sequence and create unwanted data if the test does not terminate correctly :warning: 
We recommend to use it on a duplicate to have the same modules, ...
```python3
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

import random
import string
import threading
import time

from psycopg2 import OperationalError

from odoo.service.model import PG_CONCURRENCY_ERRORS_TO_RETRY
from odoo import api, SUPERUSER_ID
from odoo.tests.common import TransactionCase


class StockQuant(TransactionCase):
    @classmethod
    def setUpClass(cls):
        super(StockQuant, cls).setUpClass()

    def test_true_concurrency_1(self):
        ran_ = random.Random('test_concurrent_update_2')
        name_prod = ''.join(ran_.choice(string.ascii_uppercase + string.digits) for _ in range(50))

        def get_cursors(nb):
            return [self.env.registry.cursor() for _ in range(nb)]

        def create_data_testing():
            with api.Environment.manage(), self.env.registry.cursor() as cursor:
                env0 = api.Environment(cursor, SUPERUSER_ID, {})
                # Create data with the env0 and commit
                stock_location_source = env0['stock.location'].create({
                    'name': 'stock_location_des',
                    'usage': 'internal',
                })
                stock_location_dest = env0['stock.location'].create({
                    'name': 'stock_location_des',
                    'usage': 'internal',
                })
                product = env0['product.product'].create({
                    'name': name_prod,
                    'type': 'product',
                })
                env0['stock.quant']._update_available_quantity(product, stock_location_source, 100_000)
                inventory = env0['stock.inventory'].create({
                    'name': 'remove product1',
                    'location_ids': [(4, stock_location_source.id)],
                    'product_ids': [(4, product.id)]
                })
                inventory.action_start()
                env0['stock.inventory.line'].search([('inventory_id', '=', inventory.id)]).product_qty = 8
                env0.cr.commit()
                env0.clear()
                return product, stock_location_source, stock_location_dest, inventory

        def clean_data_testing(product_id, inventory_id, location_source_id, location_dest_id):
            api.Environment.reset()
            with api.Environment.manage(), self.env.registry.cursor() as cursor:
                env = api.Environment(cursor, SUPERUSER_ID, {})
                product = env['product.product'].browse(product_id)
                location_source = env['stock.location'].browse(location_source_id)
                location_dest = env['stock.location'].browse(location_dest_id)
                env['stock.quant'].search([('product_id', '=', product.id)]).unlink()
                env.cr.execute("DELETE FROM stock_move WHERE product_id = %s", (product.id,))
                env.cr.execute("DELETE FROM stock_move_line WHERE product_id = %s", (product.id,))
                env.cr.execute("DELETE FROM stock_inventory_line WHERE inventory_id = %s", (inventory_id,))
                env.cr.execute("DELETE FROM stock_inventory WHERE id = %s", (inventory_id,))
                location_source.unlink()
                location_dest.unlink()
                product.unlink()
                env.cr.commit()

        def concurrent_inventory(product_id, location_id, qty):
            def todo():
                with api.Environment.manage(), self.env.registry.cursor() as cursor:
                    print(threading.current_thread().name + " : start try")
                    env = api.Environment(cursor, SUPERUSER_ID, {})

                    product = env['product.product'].browse(product_id)
                    location = env['stock.location'].browse(location_id)
                    inventory = env['stock.inventory'].create({
                        'name': 'product1_inv',
                        'location_id': location.id,
                        'product_id': product.id,
                    })
                    inventory._action_start()
                    inventory.line_ids.quantity -= qty
                    inventory._action_done()

            def try_todo(retry):
                try:
                    if retry <= 0:
                        print(threading.current_thread().name + " : Don't retry anymore")
                    else:
                        api.Environment.reset()
                        todo()
                except OperationalError as e:
                    if e.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
                        raise
                    wait_time = random.uniform(0.2, 1 ** (5 - retry))
                    print(threading.current_thread().name + f" : Wait {wait_time}")
                    time.sleep(wait_time)
                    try_todo(retry - 1)

            try_todo(5)
            print(threading.current_thread().name + " : Finished Thread")

        def concurrent_move(product_id, location_source_id, location_dest_id, qty):
            def todo():
                with api.Environment.manage(), self.env.registry.cursor() as cursor:

                    print(threading.current_thread().name + " : start try")
                    env = api.Environment(cursor, SUPERUSER_ID, {})

                    product = env['product.product'].browse(product_id)
                    location_source = env['stock.location'].browse(location_source_id)
                    location_dest = env['stock.location'].browse(location_dest_id)
                    move = env['stock.move'].create({
                        'name': 'product1_move',
                        'location_id': location_source.id,
                        'location_dest_id': location_dest.id,
                        'product_id': product.id,
                        'product_uom': product.uom_id.id,
                        'product_uom_qty': qty,
                    })
                    print(threading.current_thread().name + " : confirm move")
                    move._action_confirm()

                    print(threading.current_thread().name + " : assign move + set qty")
                    move._action_assign()
                    move.move_line_ids.qty_done = qty

                    print(threading.current_thread().name + " : move done")
                    move._action_done()

                    print(threading.current_thread().name + " : commit")
                    env.cr.commit()

            def try_todo(retry):
                try:
                    if retry <= 0:
                        print(threading.current_thread().name + " : Don't retry anymore")
                    else:
                        api.Environment.reset()
                        todo()
                except OperationalError as e:
                    if e.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
                        raise
                    wait_time = random.uniform(0.2, 1 ** (5 - retry))
                    print(threading.current_thread().name + f" : Wait {wait_time}")
                    time.sleep(wait_time)
                    try_todo(retry - 1)

            try_todo(5)
            print(threading.current_thread().name + " : Finished Thread")

        def check_result(product_id, location_source_id, location_dest_id):
            with api.Environment.manage(), self.env.registry.cursor() as cursor:
                env = api.Environment(cursor, SUPERUSER_ID, {})
                # product = env['product.product'].browse(product_id)
                location_source = env['stock.location'].browse(location_source_id)
                location_dest = env['stock.location'].browse(location_dest_id)
                # location_dest = env['stock.location'].browse(location_dest_id)
                quants = env['stock.quant'].search([('location_id', '=', location_source.id)])
                moves = env['stock.move'].search([('location_id', '=', location_source.id)])
                self.assertEqual(sum(quants.mapped("reserved_quantity")), sum(moves.mapped("reserved_availability")))

                quants_dest = env['stock.quant'].search([('location_id', '=', location_dest.id)])
                self.assertEqual(sum(quants.mapped("quantity") + quants_dest.mapped("quantity")), 100_000)
                self.assertEqual(sum(quants_dest.mapped("quantity")), sum(moves.mapped("product_uom_qty")))
                self.assertEqual(sum(quants_dest.mapped("reserved_quantity")), 0.0)

        product, stock_location_source, stock_location_dest, inventory = create_data_testing()
        try:

            pool_thread = []
            for i in range(50):  # Max connection pool is 64
                if i % 2 == 0:
                    pool_thread.append(threading.Thread(target=concurrent_inventory, name=f"T {i}", args=(product.id, stock_location_source.id, 5)))
                else:
                    pool_thread.append(threading.Thread(target=concurrent_move, name=f"T {i}", args=(product.id, stock_location_source.id, stock_location_dest.id, 5)))

            for thread in pool_thread:
                thread.start()

            for thread in pool_thread:
                thread.join()
                print("Pass a join")
                if thread.is_alive():
                    print("Timeout expired")

            print("Check result")
            check_result(product.id, stock_location_source.id, stock_location_dest.id)
        finally:
            print("Clean data")
            clean_data_testing(product.id, inventory.id, stock_location_source.id, stock_location_dest.id)

```

**Comment by randall-vx** [comment]
@moylop260 @JulioSerna do you check this comments ?

**Comment by gurneyalex** [comment]
for the record we have 3 customers at camptocamp running 12.0EE affected by this issue.

**Comment by pga-odoo** [comment]
I am not completely sure, but here are some inputs

I faced a similar issue in v14, and I think it is generated when you are using higher rounding precision values in UOM compared to global `Product Unit of Measure` in **Decimal Accuracy**. 

I don't know the steps but, With this somehow database get wrong values. According to UI, I had a `stock.quant` record with quantity 0.24 and reserved quantity 0.24 but in the database values were like this, quantity 0.239999999999999 and reserved quantity 0.23999

So now when you try to unreserve related delivery with 0.24 quantity you get this error.

**Comment by clxadmin** [comment]
Hello Experts
does anyone have an idea how the problem can be solved?

**Comment by gurneyalex** [comment]
@pga-odoo this is a different issue

**Comment by gurneyalex** [comment]
@clxadmin there are two issues:

1. the bug corrupts your current stock level
2. the bug blocks your operations

We are currently working around this with two tools, in a module (that was originally written by Forgeflow and we expanded on it to handle more cases) which
1. provides some server actions / scheduled actions to fix the stock to help create a stock take which will get the correct stock levels
2. change the method on _update_reserved_quantity to avoid blocking the operations

I can publish this code if it helps others facing the issue, but I would much rather have a fix by Odoo for the underlying issue, because this is at best a hackish workaround.

**Comment by gurneyalex** [comment]
is https://github.com/odoo/odoo/pull/65044 related to this issue, or is it something different?

**Comment by theoo** [comment]
This might be unrelated, but running out of option I tried to remove `stock.picking` manually through the shell (and start again from a clean slate).

Running the following didn't work and raises as expected the error in this issue's title. Just like with the GUI.
```
self.env['stock.picking'].browse(9732).unlink()
```

This raised the same error, when trying to simply set the _initial demand_ to zero.
```
self.env['stock.picking'].browse(9732).move_line_ids.write({'product_uom_qty': 0})
```

But, because I wasn't sure at first with attribute holds the _initial demand_, I tried to reset `product_qty` instead:
```
self.env['stock.picking'].browse(9732).move_line_ids.write({'product_qty': 0})
odoo.exceptions.UserError: ('The requested operation cannot be processed because of a programming error setting the `product_qty` field instead of the `product_uom_qty`.', '')
```
Which provides a clear complain.

**Then it becomes strange** because I was able to set the _initial demand_ to zero using the previously attribute! Then unlink the `stock.picking` and commit my changes. 
```
self.env['stock.picking'].browse(9732).move_line_ids.write({'product_uom_qty': 0})
self.env['stock.picking'].browse(9732).unlink()
self.env.cr.commit()
```
In conclusion, if I set `product_qty` to `0` first, then setting `production_uom_qty` works without validation error and I can fix this extremely annoying issue, where the solution from @amoyaux doesn't worked for me (see https://www.odoo.com/it_IT/forum/help-1/it-is-not-possible-to-unreserve-more-products-of-than-you-have-in-stock-138783 on top of his comment).

Maybe this can help.

**Comment by okuryan** [comment]
Hi, can it be that his is because of concurrent update and this savepoint code?
https://github.com/odoo/odoo/blob/13.0/addons/stock/models/stock_move.py#L1135

**Comment by yelizariev** [comment]
Better make a link this way:
https://github.com/odoo/odoo/blob/9a129ca95d80b12f6cf4bf65ff3b7fe4d0ac8cc7/addons/stock/models/stock_move.py#L1134-L1142

**Comment by mbarreto17** [comment]
> I am not completely sure, but here are some inputs
> 
> I faced a similar issue in v14, and I think it is generated when you are using higher rounding precision values in UOM compared to global `Product Unit of Measure` in **Decimal Accuracy**.
> 
> I don't know the steps but, With this somehow database get wrong values. According to UI, I had a `stock.quant` record with quantity 0.24 and reserved quantity 0.24 but in the database values were like this, quantity 0.239999999999999 and reserved quantity 0.23999
> 
> So now when you try to unreserve related delivery with 0.24 quantity you get this error.

Hola, y cómo lo corregiste?

**Comment by CasVissers-360ERP** [comment]
Did someone write a script to reset reserved quantity on stock.quant by removing current stock.move.lines completely for all products? Like a "hard reset"? I think it can only be done with job queue in bigger environments. 

This bug is still causing issues in one of our clients v13 EE environment.

**Comment by robinhli** [comment]
same issue on 13.0.

**Comment by randall-vx** [comment]
Same issue in v13.0

**Comment by thetrebelcc** [comment]
For any else going crazy like me check your reserved quantity on the back end. On the front end, it was set to 0, however, after using the XML-RPC API I scraped the back end and found the reserved quantity off by so much it was in negative scientific notation? Never seen anything like this on Odoo. I wish I had taken a screenshot. Admittedly this was probably caused by me, I was trying to move stuff around to force the MO closed. I reset all the qtys, then matched it with the reserved, was finally able to close the MO after many hours.  

My thoughts are

-Somewhere in the large number of moves the onhand and reserved got out of synch. 

- It seems pickings are what causes the out-of-synch(for us). 

-Moves are being allowed to happen when the res/on hand are out of synch(note: backorders are NOT created)

- Unable to close records, because of reservation issues. Almost impossible to track the root move that caused the issue if you are running a large number of moves. 


The above solutions are nice, but since we are all running different versions of UOMs /decimal precisions and routes it's going to be extremely difficult to find a "one size fits all" solution. 

 My python3 code is below, running the latest version of Odoo 13 enterprise. Variables are your own. Also make sure your location has at least 1, if set to zero it will not be picked up for some reason. I understand this in theory. none in room == doesn't exist. It would seem better to still be able to get the ID. 


```
on_hand_data = models.execute_kw(db, uid, password,
                                 'stock.quant', 'search_read',
                                 [[['product_id', '=', product],
                                   ['location_id', '=',location ]
                                   ]])
### check reserved and other qtys for strange numbers or other anomalies, again these are NOT apparent on the front end. 

fix_rsv = models.execute_kw(db, uid, password, 'stock.quant', 'write',
                      [on_hand_data[0]['id', {
                          'reserved_quantity': int(fix_qty)
                      }])
```

**Comment by thetrebelcc** [comment]
If anyone like me, is desperate to close the record I made a quick script to basically Bruteforce the record close. Not an elegant solution but it gets the job done. 

https://github.com/thetrebelcc/Odoo-Unable-to-Unreserve-Fix-Script

**Comment by victor-champonnois** [comment]
@amoyaux Do you know if this bug is fixed in Odoo 16 ?

**Comment by wild3r** [comment]
In my case I solved it as follows:
step 01: Update the product module
step 02: I validated again and everything was ok

**Comment by bfaunce-aml** [comment]
We are still having this issue on our deployment (v17), possible to replicate but not consistently. Is there any timeline for looking into this?

**Comment by niecw** [comment]
We've experience this issue but caused by the figure not being rounded properly.  This affect, AFAIK, versions v11 till the latest (not tested, just by the checking for the code similar to the line 221 of the snippet mentioned in the comment https://github.com/odoo/odoo/issues/62139#issuecomment-734342663 above.  Here is the evidence:

<img width="400" height="232" alt="Image" src="https://github.com/user-attachments/assets/1a972f51-8b39-48cc-9e8c-de104c96b8d6" />

The behaviour caused the stock to be "slightly" under (or sometimes slightly over).  The following figures show the sampe data we have experienced in one instance, with the SQL to list the occurences:

<img width="658" height="524" alt="Image" src="https://github.com/user-attachments/assets/4e7e4100-fe35-4c6e-a1f2-b07282bb8458" />

**Comment by bfaunce-aml** [comment]
@niecw I think that this thread is addressing two separate issues:

1. The "Not possible to unreserve more products than you have in stock" error can be triggered because of a minor rounding error (as you just mentioned)
2. The error can be triggered because of a bug in `_update_available_quantity` which causes twice as much inventory as intended to be taken out of stock, leading to a negative reserved quantity on the `stock.quant`

I'm facing the second issue here. Have you seen anything that might indicate a solution to this?

**Comment by niecw** [comment]
> I'm facing the second issue here. Have you seen anything that might indicate a solution to this?

(Un)fortunately I have not experience the second issue.  The first issue is also within the `_update_available_quantity` though I beleive of different nature.
